cardinal/ecs/cmdbuffer/cmdbuffer.go

```diff
@@ -0,0 +1,508 @@
+package cmdbuffer
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/redis/go-redis/v9"
+	"pkg.world.dev/world-engine/cardinal/ecs/archetype"
+	"pkg.world.dev/world-engine/cardinal/ecs/codec"
+	"pkg.world.dev/world-engine/cardinal/ecs/component"
+	"pkg.world.dev/world-engine/cardinal/ecs/entity"
+	"pkg.world.dev/world-engine/cardinal/ecs/filter"
+	"pkg.world.dev/world-engine/cardinal/ecs/log"
+	"pkg.world.dev/world-engine/cardinal/ecs/storage"
+	"pkg.world.dev/world-engine/cardinal/ecs/store"
+)
+
+var _ store.IManager = &Manager{}
+
+type Manager struct {
+	client *redis.Client
+
+	compValues         map[compKey]any
+	compValuesToDelete map[compKey]bool
+	typeToComponent    map[component.TypeID]component.IComponentType
+
+	activeEntities map[archetype.ID]activeEntities
+
+	// Fields that track the next valid entity ID that can be assigned
+	nextEntityIDSaved uint64
+	pendingEntityIDs  uint64
+	isEntityIDLoaded  bool
+
+	// Archetype ID management.
+	entityIDToArchID       map[entity.ID]archetype.ID
+	entityIDToOriginArchID map[entity.ID]archetype.ID
+
+	archIDToComps  map[archetype.ID][]component.IComponentType
+	pendingArchIDs []archetype.ID
+
+	logger *log.Logger
+}
+
+var (
+	errorArchIDNotFound     = errors.New("archetype for components not found")
+	doesNotExistArchetypeID = archetype.ID(-1)
+)
+
+// NewManager creates a new command buffer manager that is able to queue up a series of states changes and
+// atomically commit them to the underlying redis storage layer.
+func NewManager(client *redis.Client) (*Manager, error) {
+	m := &Manager{
+		client:             client,
+		compValues:         map[compKey]any{},
+		compValuesToDelete: map[compKey]bool{},
+
+		activeEntities: map[archetype.ID]activeEntities{},
+
+		entityIDToArchID:       map[entity.ID]archetype.ID{},
+		entityIDToOriginArchID: map[entity.ID]archetype.ID{},
+
+		// These fields cannot be set until RegisterComponents is called
+		archIDToComps:   nil,
```

Comment:
Put a boolean check here.  

Or if possible make it so that NewManager has to instantiate with components. The latter is the safest bet as it becomes a static check but may make the logic a feel more awkward as we now require components to exist before the buffer.

Reply:
Unfortunately, we won't have a full list of component at this point. This object needs to be passed into the ecs.World, and components can only be registered after the world object has been created.